### PR TITLE
feat: use-sync-external-store

### PR DESCRIPTION
**Pull request by dai-shi**
## No public API changes

The API is compatible with the latest v3 API

## Major changes

- React part is fully re-implemented with [use-sync-external-store](https://www.npmjs.com/package/use-sync-external-store)

## Better support for context usage

- It exports a new `useStore` hook.
- It re-exports `createStore` from `zustand/vanilla`
- `zustand/context` is fully rewritten with the new `useStore`.

### CODE REVIEW

**Comment by dai-shi** [comment]
Four tests failed:
- re-renders with useLayoutEffect
- can update the equality checker
- can throw an error in selector
- can throw an error in equality checker

I feel like we should hit v4 major with breaking changes.

**Comment by dai-shi** [comment]
I hacked around, but not sure if this causes a problem with cases where we used to have zombie children.

**Comment by dai-shi** [comment]
https://www.npmjs.com/package/zustand/v/4.0.0-alpha.2

File: tests/basic.test.tsx [hunk 1]
```diff
@@ -150,7 +150,9 @@ it('re-renders with useLayoutEffect', async () => {
 
   const container = document.createElement('div')
   ReactDOM.render(<Component />, container)
-  expect(container.innerHTML).toBe('true')
+  waitFor(() => {
```
Comment (timkindberg) [inline tests/basic.test.tsx:153]: I believe you need an `await` in front?

**Comment by dai-shi** [comment]
meant `4` 😓 
https://www.npmjs.com/package/zustand/v/4.0.0-alpha.4

**Comment by dai-shi** [comment]
https://www.npmjs.com/package/zustand/v/4.0.0-alpha.6

**Comment by dai-shi** [comment]
https://www.npmjs.com/package/zustand/v/4.0.0-alpha.7

**Comment by dai-shi** [comment]
@phaistonian Yes. Please see #971.

**Comment by phaistonian** [comment]
@dai-shi  Awesome!

Any chance this change is the _culprit_ for [this](https://github.com/pmndrs/zustand/issues/972)?

**Pull request merged**